ultimate-pipeline/src/ultimate_pipeline/pipelines/data_processing/supervisely_downloader.py

```diff
@@ -34,14 +38,43 @@ def helper_download_image_dataset_from_supervisely(params: t.Dict) -> t.Tuple[t.
     annotations_dict = {}
     images_dict = {}
 
-    for image in tqdm(image_info_list):
+    for image in image_info_list:
 
         image_name = Path(image.name).stem
 
-        img_data = Image.fromarray(api.image.download_np(image.id))
-        images_dict[image_name] = img_data
-
-        image_ann_json = api.annotation.download(image.id).annotation
-        annotations_dict[image.name] = image_ann_json
+        images_dict[image_name] = download_image_lazy(image, api)
```

Comment:
TBH, even with annotations, it's not easy to spot that `download_image_lazy` and `download_annotation_lazy` are high-order functions (i.e. functions accepting or returning other functions).

IMHO, a shorter and more idiomatic code would be to define `download_image` and `download_annotation` as local functions of 1 param (which access `api` via closure), and then in the loop put them in lambda expressions. This will effectively convert the local functions of 1 parameter to functions of 0 parameters returning np.ndarray. Using lambda is also more idiomatic and emphasizes that what we put in dictionaries are functions (generating content).
```
    annotations_dict = {}
    images_dict = {}

    def download_image(img) -> np.ndarray:
        logger.info(f"Start downloading image {img.name} from Supervisely")
        return Image.fromarray(api.image.download_np(img.id))
    
    def download_annotation(img)-> t.Dict[str, t.Any]:
        logger.info(f"Start downloading annotation for image {img.name} from Supervisely")
        return api.annotation.download(img.id).annotation

    for image in image_info_list:
        image_name = Path(image.name).stem
        images_dict[image_name] = lambda: download_image(image, api)
        annotations_dict[image.name] = lambda: download_annotation(image, api)

    return metadata, annotations_dict, images_dict
```

Reply:
thanks :) than i will go for this approach

Reply:
Ha, good catch... Must be related to the way how Python scopes iteration variables (or rather, doesn't scope).
I've never come across that issue in C#, even though I used that pattern often. But in C#, the variables inside the loop are scoped properly, meaning that each iteration would get a new `image` variable, and lambda from the previous iteration would still keep a reference to the old `image` variable.
Not sure about JavaScript - possibly the old behaviour with `var` inside the loop would behave like Python, but with the newer `let` binding like C#...